man/man1/peak.1
.TH peak 1 "04 Sept 1998"
.SH NAME
peak \- peak measurements 
.SH SYNOPSIS
.B peak
[options] image.fits  
.SH DESCRIPTION
.PP
.B peak
detects localizes bright object centroids in an image. It takes as
argument a list of FITS cube names and outputs on stdout the center of
each detected bright zone, an estimation of FWHM in x and y, an average
FWHM, a flux estimation, and the minimum and maximum pixel values around
each object.
.PP
Two different detection method can be used with the 
.B peak
command: the 'kappa-sigma' method (default) or the 'squares' method (use
the -m (--method) option to choose).
.PP
With the 'kappa-sigma' method, bright objects are detected if and only if 
they contain a peak of sufficient amplitude i.e. more than K deviations above 
the median image value, AND if they cover a surface at least of 3x3 pixels. 
To detect peaks located in a window smaller than 3x3, there is an option to 
smear out the image by a low-pass filter before applying peak detection, but
of course it has associated drawbacks.
By default,
.B peak
will be looking at all the signal which is more than 2 deviations above
the median pixel value. This value can be changed using the \-k (or
\--kappa) option.
Notice that this is not more than kappa-sigma clipping for signal
detection, except that to avoid mis-estimation of the mean and sigma in
crowded fields, the measurements are done with the median and a
deviation which is the average absolute distance to the median.
.PP
In the 'square' method, a standard deviation filter is apply to the image,
which has the effect to make the bright points appear like squares. This
squares image is then used as a mask to detect objects.
.PP
In spite of calls to morphological filters, 
.B peak 
is surprisingly faster than other usual algorithms. It should be used
as a peak position estimator more than a precise locator, though. 
.SH ALGORITHMS
.PP
.B Kappa-sigma method
.PP
If smearing is activated, the input image is smeared out with a low-pass
5x5 filter before detection is applied. This filtered version of the
input image is only used for detection purposes, and not for any kind of
later measurement.
.PP
A binary map is first created of all pixel positions which have a value
above a given threshold (by default, median plus 2 deviations).
.PP
A binary morphological erosion, and a dilation are then performed
on the binary map to close all regions smaller than 3x3, which removes
all isolated bad pixels. If smearing was applied, small objects would
have been enlarged to bigger than 3x3 and appear in the resulting pixel
map.
.PP
A floodfill algorithm is applied to find the geometric center of
all white blobs, weighted by pixel values taken from the original image.
.PP
.B Squares method
.PP
A standard deviation filter is applied to the image which make the bright
objects appaear like bright squares. This bright squares image is then
binarized and used as a mask to identify zones where the objects are in the
original image. A morphological closing is then applied, and the remaining
objects are registered.
.PP
For both methods, if fine positioning is activated (\-f or \--finepos option), 
a subsequent process is called, which requests 3 user-defined radiuses in
pixels: r1, r2, and r3. For each found peak position in the image, a
background is computed as the median pixel value in the ring centered on
the estimated peak position, of radiuses r2 and r3. Then a barycenter is
computed within the disk centered on the same spot, of radius r1, using
background-subtracted pixel values as weights. This fine positioning
method proves quite reliable, but requires all peaks in the image
to have more or less the same size to fit into the circles defined by
r1, r2, r3.
.SH OPTIONS
.TP
.BI \-m " "clip " or " \--method " "clip
Use 'kappa-sigma' detection method.
.TP
.BI \-m " "squares " or " \--method " "squares
Use 'squares' detection method.
.TP
.BI \-k " cut " or " " \--kappa " cut"
To be used for 'kappa-sigma' method.
Use this option to change the cut level in a factor of deviations. The
lower this value, the more bright zones may be detected. The higher this
factor is, the less detected peaks. The default of 2.0 seems to work
fine on images having a high Signal to Noise Ratios.
.TP
.B \-s " or " \--smear
This option (low-pass filter) applies a 5x5 convolution with
a flat kernel before trying to detect objects. The smearing is
h.PPful to detect objects which are smaller than a 3x3 window.
It increases the number of detections, but also the number of
false detections. Bad pixels, for example, are smeared out to
a 5x5 window and detected as proper peaks. Another issue
is that 2 close peaks will be smeared out to a single one. Most
probably, the returned result will be a barycenter of the 2 regions
instead of the 2 expected centers. Because the smearing will lower the
signal in all regions, the default sigma cut is halved when this options
is used.
.TP
.BI \-S " or " \--sqhsize " 'hx hy'"
To be used for 'squares' method.
Define the size of the standard deviation filter applied to generate the
squares image. The bigger the filter is, the bigger the squares are.
.TP
.BI \-f " 'r1 r2 r3' " or " " \--fpos " 'r1 r2 r3'"
Fine positioning: provide three values r1 < r2 < r3.
The radiuses r2 and r3 specify a ring around each
detected point, from which an estimation of the background is computed.
A barycenter is then computed in the disk of radius r1, using
background-subtracted pixel values as weights. No defaults are given
to these parameters.
.br
Be aware when using this position refining that it assumes the
following conditions.
.PP
All peaks are isolated, i.e. the closest distance between 2 peaks is
strictly greater than 2 * r3.
.PP
There is a background zone around every peak, always within the disk
defined by r2 and r3. Otherwise, the background estimation is contaminated
with peak signals.
.PP
All peaks are contained in a disk of radius r1.
.TP
.BI \-F " or " \--fwhm 
Flag to print out the fhwm for detected objects.
.TP
.BI \-P " or " \--phot " 'r1 r2 r3'"
Provides the radiuses used to compute photometry.
.TP
.BI \-d " or " \--rtd 
Flag to display detected objects in rtd.
.SH BUGS
.PP
Peaks located on an image edge will not be detected.


CHANGES
List of changes between eclipse 4.9 and eclipse 5.0
---------------------------------------------------

Since last delivery, there has been no major channge in the library.
Only a couple of minor fixes were brought to the ISAAC pipeline.



src/main/peak.c
/*----------------------------------------------------------------------------*/
/**
   @file    peak.c
   @author  N. Devillard
   @date    June 2001
   @version	$Revision: 1.39 $
   @brief   Astronomical object detector and stat computation
*/  
/*----------------------------------------------------------------------------*/

/*
	$Id: peak.c,v 1.39 2003/08/07 09:29:58 yjung Exp $
	$Author: yjung $
	$Date: 2003/08/07 09:29:58 $
	$Revision: 1.39 $
*/

/*------------------------------------------------------------------------------
                                Defines
 -----------------------------------------------------------------------------*/

#define OPT_METHOD		            1000
#define OPT_KAPPA		            1010
#define OPT_SMEAR		            1011
#define OPT_SQHSZ		            1020

#define OPT_FINEPOS		            1030
#define OPT_FWHM		            1031
#define OPT_PHOT	                1032

#define OPT_RTD 		            1040

#define DETECT_UNSET				0
#define DETECT_WITH_KAPPASIGMA		1
#define DETECT_WITH_SQUARES			2

/*------------------------------------------------------------------------------
                                Includes
 -----------------------------------------------------------------------------*/

#include <stdio.h>
#include <string.h>
#include "eclipse.h"

/*------------------------------------------------------------------------------
                                Function prototypes
 -----------------------------------------------------------------------------*/

static void usage(char *pname) ;
static char prog_desc[] = "object detection and stat computation" ;

/*------------------------------------------------------------------------------
                                    Main 
 -----------------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
    int             c ;
	int		        ret ;
	int		        detect_method ;
	int		        fpos_flag ;
	double	        fpos_star, fpos_int, fpos_ext ;
	int		        pos_only ;
	int		        sq_hx, sq_hy ;
	double	        kappa ;
	int		        smear_flag ;
	int		        fwhm_flag ;
	int		        phot_flag ;
	double	        phot_star, phot_int, phot_ext ;
	int		        hx, hy ;
	int		        rtd_active ;
	detected	*	det ;
	double3		*	pts ;
	cube_t		*	c_in ;
	int				i, p ;
	int				err ;
	
    if (argc<2) usage(argv[0]);

    /* Initialize */
	detect_method = DETECT_WITH_KAPPASIGMA ;
	hx 			= -1 ;
	hy			= -1 ;
	smear_flag	= 0 ;
	fwhm_flag	= 0 ;
	fpos_flag 	= 0 ;
	fpos_star	= DETECTED_FPOS_STAR ;
	fpos_int	= DETECTED_FPOS_INT ;
	fpos_ext	= DETECTED_FPOS_EXT ;
	pos_only	= 0 ;
	sq_hx		= DETECTED_SQHX ;
	sq_hy		= DETECTED_SQHY ;
	kappa		= DETECTED_KAPPA ;
	phot_flag	= 0 ;
	phot_star	= DETECTED_PHOT_STAR ;
	phot_int	= DETECTED_PHOT_INT ;
	phot_ext	= DETECTED_PHOT_EXT ;
	rtd_active 	= 0 ;

    while (1) {
        int     option_index = 0 ;
        static struct option long_options[] =
        {
            {"license", 0, 0, OPT_LICENSE},
            {"help",    0, 0, OPT_HELP},
            {"version", 0, 0, OPT_VERSION},

			{"method",	1, 0, OPT_METHOD},
			{"kappa",	1, 0, OPT_KAPPA},
			{"smear",	0, 0, OPT_SMEAR},
			{"sqhsize",	1, 0, OPT_SQHSZ},

			{"fpos",	1, 0, OPT_FINEPOS},
			{"fwhm",	0, 0, OPT_FWHM},
			{"phot",	1, 0, OPT_PHOT},

			{"rtd",		0, 0, OPT_RTD},

            {0, 0, 0, 0}

        } ;
        c = getopt_long(argc,
                        argv,
                        "df:Fhk:m:P:sS:L",
                        long_options,
                        &option_index) ;
        if (c==-1) break ;

        switch(c) {
            /* Standard option: display license undocumented option */
            case OPT_LICENSE:
            case 'L':
                eclipse_display_license() ;
                return 0 ;

            /* Standard option : help */
            case OPT_HELP:
            case 'h':
                usage(argv[0]) ;
                break ;

            /* Standard option: version */
            case OPT_VERSION:
                print_eclipse_version() ;
                return 0 ;

            /* Local options */
			case OPT_METHOD:
			case 'm':
                if (!strcmp(optarg, "clip")) {
                    detect_method = DETECT_WITH_KAPPASIGMA ;
                } else if (!strcmp(optarg, "squares")) {
                    detect_method = DETECT_WITH_SQUARES ;
                } else {
                    e_error("invalid method name: [%s]", optarg);
                    return -1 ;
                }
                break ;

			case OPT_KAPPA:
			case 'k':
                ret = sscanf(optarg, "%lg", &kappa);
                if (ret!=1) {
                    e_error("-k/--kappa expects 1 argument, received %d", ret);
                    return -1 ;
                }
                break ;
			case OPT_SMEAR:
			case 's':
                smear_flag=1 ;
                break ;
			case OPT_SQHSZ:
			case 'S':
                ret = sscanf(optarg, "%d %d", &hx, &hy);
                if (ret!=2) {
                    e_error("-S/--sqhsize expects 2 arguments, not %d", ret);
                    return -1 ;
                }
                break ;
			case OPT_FINEPOS:
			case 'f':
                fpos_flag=1 ;
                ret=sscanf(optarg,"%lg %lg %lg",&fpos_star,&fpos_int,&fpos_ext);
                if (ret!=3) {
                    e_error("-f/--fpos expects 3 arguments, received %d", ret);
                    return -1 ;
                }
                break ;
			case OPT_FWHM:
			case 'F':
                fwhm_flag=1 ;
                break ;
			case OPT_PHOT:
			case 'P':
                phot_flag=1;
                ret=sscanf(optarg,"%lg %lg %lg",&phot_star,&phot_int,&phot_ext);
                if (ret!=3) {
                    e_error("-P/--phot expects 3 arguments, received %d", ret);
                    return -1 ;
                }
                break ;
			case OPT_RTD:
            case 'd':
                rtd_active=1 ;
                break ;
            default:
                usage(argv[0]) ;
                break ;
        }
    }

    /* Initialize eclipse environment */
    eclipse_init() ;

    if ((argc-optind) < 1) {
        e_error("missing arguments: input file name") ;
        return -1 ;
    }
	while ((argc-optind)>0) {
		/* Load input cube */
		if ((c_in = cube_load(argv[optind])) == NULL) {
			e_error("loading %s: aborting", argv[optind]);
			return -1 ;
		}

		printf("# file: %s\n", argv[optind]);
		for (p=0 ; p<c_in->np ; p++) {
			if (rtd_active) {
				e_comment(0, "displaying image...");
				rtd_image_put(c_in->plane[p]);
			}
			/* Detect positions and fill up detected object */
            det = NULL ;
			switch(detect_method) {
				case DETECT_WITH_KAPPASIGMA:
                    det = detected_ks_engine(c_in->plane[p], kappa, smear_flag);
                    break ;
				case DETECT_WITH_SQUARES:
                    det = detected_sq_engine(c_in->plane[p], hx, hy) ;
                    break ;
                default:
                    e_error("no method specified for detection");
                    break ;
			}
			if (det == NULL) {
				e_error("in detection on plane %d: aborting", p+1);
				cube_del(c_in);
				return -1 ;
			}
			if (det->nbobj<1) {
				e_comment(0, "no object found in plane %d of [%s]", p+1,
							argv[optind]);
			} else {
				/* Do fine positioning if requested */
				if (fpos_flag) {
					err+= detected_compute_finepos(	det,
													c_in->plane[p],
													fpos_star,
													fpos_int,
													fpos_ext);
				}
				/* Do FWHM computation if requested */
				if (fwhm_flag) err+= detected_compute_fwhm(det, c_in->plane[p]);
				/* Do Photometry computation if requested */
				if (phot_flag) {
					err+= detected_compute_phot(det,
												c_in->plane[p],
												phot_star,
												phot_int,
												phot_ext);
				}
				/* Display results if requested */
				if (rtd_active) {
					pts = double3_new(det->nbobj);
					for (i=0 ; i<pts->n ; i++) {
						pts->x[i] = det->x[i] ;
						pts->y[i] = det->y[i] ;
						pts->z[i] = 0 ;
					}
					e_comment(0, "displaying found objects...");
					rtd_point_plot(pts);
					double3_del(pts);
				}
			}
			detected_dump(det, stdout);
			detected_del(det);
		}
		cube_del(c_in);
		optind++ ;
	}
	return 0 ;
}

static void usage(char * pname)
{
    hello_world(pname, prog_desc) ;
    printf("use : %s [options] in\n", pname) ;
	printf(
	"options are:\n"
	"\t-s (--smear)              Smear image before detection\n"
	"\t-f (--fpos) 'r1 r2 r3'    Set radiuses for fine positioning\n"
	"\t-F (--fwhm)               Print out FWHM for all objects\n"
	"\t-P (--phot) 'r1 r2 r3'    Compute photometry for all objects\n"
	"\t-d (--rtd)                Display image and results on RTD\n");
	printf(
	"\n"
	"\t-m (--method) clip        Use kappa-sigma clipping\n"
	"\t-k (--kappa) value        Set value for kappa-sigma clipping\n"
	"\n"
	"\t-m (--method) squares     Use squares method (experimental)\n"
	"\t-S (--sqhsize) size       Set square size\n"
	);
    printf("\n\n") ;
    exit(1) ;
}

